Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -136,16 +136,60 @@ public final class PlayerInputSystem: System {
       if event.key == .enter {
         if !message.isEmpty {
           try connection?.sendPacket(ChatMessageServerboundPacket(message))
+          guiState.playerMessageHistory.append(message)
+          guiState.currentMessageIndex = nil
         }
         guiState.messageInput = nil
         return true
       } else if event.key == .escape {
         guiState.messageInput = nil
+        guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          // If there are no messages, do nothing
+          if guiState.playerMessageHistory.count > 0 {
+            // Index up a message
+            guiState.currentMessageIndex = guiState.playerMessageHistory.count - 1
+            if let currentMessageIndex: Int = guiState.currentMessageIndex {
```

Comment:
The type annotation is unnecessary, and since it is simple it should be left out.